FirstVoicesData/src/main/java/ca/firstvoices/listeners/FVDocumentListener.java

```
@@ -183,23 +183,46 @@ public void validateCharacter() {
         && !document.isVersion()) {
       try {
         DocumentModelList characters = getCharacters(document);
+        DocumentModel alphabet = getAlphabet(document);
 
         if (event.getName().equals(DocumentEventTypes.BEFORE_DOC_UPDATE)) {
-          List<DocumentModel> results = characters.stream()
-              .map(c -> c.getId().equals(document.getId()) ? document : c)
+          //All character documents except for the modified doc
+          List<DocumentModel> filteredCharacters = characters.stream()
+              .filter(c -> !c.getId().equals(document.getId()))
               .collect(Collectors.toList());
-          cleanupCharactersService.mapAndValidateConfusableCharacters(results);
+          cleanupCharactersService.validateCharacters(filteredCharacters, alphabet, document);
         }
 
         if (event.getName().equals(DocumentEventTypes.ABOUT_TO_CREATE)) {
-          cleanupCharactersService.mapAndValidateConfusableCharacters(characters);
+          cleanupCharactersService.validateCharacters(characters, alphabet, document);
+
         }
 
       } catch (Exception exception) {
         rollBackEvent(event);
         throw exception;
       }
     }
+
+    //If doc is alphabet, do another operation for ignored characters
+    if (document.getDocumentType().getName().equals(FV_ALPHABET) && !document.isProxy()
+        && !document.isVersion()) {
+      try {
+
+        //only test on update, not creation as characters will not exist during creation
+        if (event.getName().equals(DocumentEventTypes.BEFORE_DOC_UPDATE)) {
+          DocumentModelList characters = getCharacters(document);
+          DocumentModel alphabet = getAlphabet(document);
+          cleanupCharactersService.validateAlphabetIgnoredCharacters(characters, alphabet);
+
+        }
+
+      } catch (Exception exception) {
+        rollBackEvent(event);
```

Curious if there a particular reason why we're rolling back the event here (but not above)?

---

The only reson for the roll back event here is that the character updates and in a similar try-catch block. Unsure if it needs to change or not.

---

I don't think it's entirely necessary. On error you can just throw an exception and it will bubble up and be detected by the Nuxeo transaction layer which will roll back the current transaction.